### PR TITLE
feat(web-components): add min/maxCharacters to ic-text-field

### PR DESCRIPTION
**Pull request by gd2910**
provide native functionailty to ic-text-field with min-characters and max-characters props, includes warnings for exceeding and aria labels

## Summary of the changes
provide native functionality to ic-text-field with min-characters and max-characters props, includes warnings for exceeding and aria labels. Validation status for each, spec tests added for new methods.

## Related issue
#604 

## Checklist
- [x] Relevant unit tests and visual regression tests added. 
- [x] Visual testing against Figma component specification completed. 
- [x] All acceptance criteria reviewed and met. 
- [x] Accessibility Insights FastPass performed.
- [x] A11y unit test added and yields no issues.
- [x] A11y plug-in on Storybook yields no issues. 
- [x] Manual screen reader testing performed using NVDA and VoiceOver. 
- [x] Page can be zoomed to 400% with no loss of content. 
- [x] Screen magnifier used with no issues. 
- [x] Text resized to 200% with no loss of content.
- [x] Text spacing increased as per the [WCAG 1.4.12 success criterion](https://www.w3.org/TR/WCAG21/#text-spacing) with no loss of content.
- [x] Browser setting 'prefers reduced motion' tested. No animations or motion visible whilst this setting is on. 
- [x] Windows High Contrast mode tested with no loss of content.
- [x] System light and dark mode tested with no loss of content. 
- [x] Manual keyboard testing for keyboard controls and logical focus order. 
- [x] Min/max content examples tested with no loss of content or overflow. 
- [x] Browser support tested (Chrome, Safari, Firefox and Edge).
- [x] Correct roles used and ARIA attributes used correctly where required. 
- [x] Logical heading structure is maintained, and the HTML elements used for headings can be changed to fit within the wider page structure. 
- [x] All prop combinations work without issue. 
- [x] Changes to docs package checked and committed.

### CODE REVIEW

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 1]
```diff
@@ -372,13 +388,32 @@ export class TextField {
...
+        this.numChars >= this.maxCharacters ? true : false;
+    }
+  };
+
+  private getMinCharactersUnattained = (value: string) => {
+    this.numChars = value.length;
+
+    if (this.minCharacters > 0) {
+      this.minCharactersUnattained =
+        this.numChars < this.minCharacters ? true : false;
```
Comment (GCHQ-Developer-741) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:405]:
```suggestion
        this.numChars < this.minCharacters;
```

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 2]
```diff
@@ -372,13 +388,32 @@ export class TextField {
     }
   };
 
+  private getmaxCharactersReached = (value: string) => {
+    this.numChars = value.length;
+
+    if (this.maxCharacters > 0) {
+      this.maxCharactersReached =
+        this.numChars >= this.maxCharacters ? true : false;
```
Comment (GCHQ-Developer-741) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:396]:
```suggestion
        this.numChars >= this.maxCharacters;
```

**Review by GCHQ-Developer-741 — APPROVED** [review]

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:1]: Should the `maxCharacters` prop prevent the user from actually typing any more characters than the maximum? A bit like the [HTML text field](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/maxlength). It sounds like that might be the required behaviour looking at the ticket (not 100% sure though) - but then if that is the case, the maximum can't be exceeded so not sure when the error would appear

**Review by GCHQ-Developer-847 — CHANGES_REQUESTED** [review]
A few comments :)

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx
Comment (GCHQ-Developer-847) [reply packages/web-components/src/components/ic-text-field/ic-text-field.tsx:1]: As discussed, ignore the first part of this - it does work, sorry 🤦‍♀️ There's just the issue of when the error should appear (if at all)

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 3]
```diff
@@ -158,13 +161,23 @@ export class TextField {
   /**
    * The maximum number of characters that can be entered in the field.
    */
+  @Prop() maxCharacters: number = 0;
+
+  /**
+   * The label for maximum number of characters that can be entered in the field.
```
Comment (GCHQ-Developer-299) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:167]: This comment for MaxLength doesn't seem to line up - how can a label be a number?

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 4]
```diff
@@ -56,6 +56,9 @@ export class TextField {
   @Element() el: HTMLIcTextFieldElement;
 
   @State() numChars: number = 0;
+  @State() maxCharactersReached: boolean = false;
+  @State() maxCharactersError: boolean = false;
```
Comment (GCHQ-Developer-741) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:60]: I'm a bit confused why these two state values appear to be doing a similar thing. They also appear to be based off one another. Is it necessary for both of these to be considered `state`?

Comment (gd2910) [reply packages/web-components/src/components/ic-text-field/ic-text-field.tsx:60]: I needed separate states to control the error state occurring and the maxlength being applied to the input, without these the error state is applied once the maxCharacters is reached, when we only want it to be applied once the user attempts to exceed the maxLength.

**Review by GCHQ-Developer-299 — DISMISSED** [review]

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 5]
```diff
@@ -345,6 +362,9 @@ export class TextField {
   @Listen("keydown", {})
   handleKeyDown(ev: KeyboardEvent): void {
     this.icKeydown.emit({ event: ev });
+    this.maxCharactersReached
```
Comment (GCHQ-Developer-741) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:365]:
```suggestion
    this.maxCharactersError = this.maxCharactersReached
```

**Review by DBD324 — APPROVED** [review]
From a functional perspective this seems to work well. Thanks

**Review by GCHQ-Developer-847 — APPROVED** [review]

**Pull request merged**